src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -76,16 +78,58 @@ double KDERules<MetricType, KernelType, TreeType>::BaseCase(
   if ((lastQueryIndex == queryIndex) && (lastReferenceIndex == referenceIndex))
     return 0.0;
 
-  // Calculations.
-  const double distance = metric.Evaluate(querySet.col(queryIndex),
-                                          referenceSet.col(referenceIndex));
-  densities(queryIndex) += kernel.Evaluate(distance);
+  // Use PCA evaluation if possible.
+  if (pca)
+  {
+    const KDEStat& stat = traversalInfo.LastReferenceNode()->Stat();
+
+    const arma::vec& qPoint = querySet.col(queryIndex);
+    const arma::vec qProj = stat.EigVec().t() * (qPoint - stat.Mean());
+    const arma::vec qRecon = stat.EigVec() * qProj + stat.Mean();
+
+    // Base case for each reference node.
+    const arma::vec& rPoint = referenceSet.col(referenceIndex);
+    const arma::vec rProj = stat.EigVec().t() * (rPoint - stat.Mean());
+    const arma::vec rRecon = stat.EigVec() * rProj + stat.Mean();
```

Comment:
As far as I can tell, these computations are also correct assuming that `stat.EigVec()` holds one eigenvector in each column.  (I am not sure if that is true... I think it is, but maybe it is worth double-checking.)